### PR TITLE
Telemetry policy

### PR DESCRIPTION
**Pull request by antkmsft**
Closes https://github.com/Azure/azure-sdk-for-cpp/issues/253

### CODE REVIEW

File: sdk/core/azure-core/inc/http/policy.hpp [hunk 1]
```diff
@@ -117,4 +125,37 @@ namespace Azure { namespace Core { namespace Http {
...
+        std::string const& componentVersion,
+        std::string const& applicationId);
+
+  public:
+    explicit TelemetryPolicy(std::string const& componentName, std::string const& componentVersion)
+        : TelemetryPolicy(componentName, componentVersion, g_emptyApplicationId)
+    {
+    }
+
+    explicit TelemetryPolicy(
```
Comment (JeffreyRichter) [inline sdk/core/azure-core/inc/http/policy.hpp:144]: Can you make this work more like this: https://github.com/Azure/azure-storage-blob-go/blob/master/azblob/zc_policy_telemetry.go

Specifically, the ctor should take 1 options parameter which has 1 field which shoul dbe call application_id to match this guidelines: https://azure.github.io/azure-sdk/general_azurecore.html (the go code calls this value but the go code should change to application_id)

Then, the rest of the string should be created by the policy itself so that all callers don't have to redo the same work all the time and this also ensures that the string is correct and consistent when sent to the the service.

File: sdk/storage/inc/common/common_headers_request_policy.hpp [hunk 2]
```diff
@@ -12,7 +12,10 @@ namespace Azure { namespace Storage {
     explicit CommonHeadersRequestPolicy() {}
     ~CommonHeadersRequestPolicy() override {}
 
-    HttpPolicy* Clone() const override { return new CommonHeadersRequestPolicy(*this); }
+    std::unique_ptr<HttpPolicy> Clone() const override
```
Comment (JeffreyRichter) [inline sdk/storage/inc/common/common_headers_request_policy.hpp:15]: What does this CommonHeadersRequestPolicy class even do? 
There is no comment explaining it. Why do we need this class?

Comment (JinmingHu-MSFT) [reply sdk/storage/inc/common/common_headers_request_policy.hpp:15]: It's used to add headers that apply to every HTTP request. But for the moment, I guess there's no other headers besides `x-ms-date`. So I guess we should rename it to `DatePolicy`?

File: sdk/core/azure-core/inc/credentials/policy/policies.hpp [hunk 3]
```diff
@@ -24,40 +24,33 @@ namespace Azure { namespace Core { namespace Credentials { namespace Policy {
     void operator=(BearerTokenAuthenticationPolicy const&) = delete;
 
   public:
-    BearerTokenAuthenticationPolicy(
-        std::shared_ptr<TokenCredential const> const& credential,
-        std::string const& scope)
-        : m_credential(credential)
+    explicit BearerTokenAuthenticationPolicy(
```
Comment (RickWinter) [inline sdk/core/azure-core/inc/credentials/policy/policies.hpp:27]: Nit, This looks unrelated to the telemetry policy feature.  Easier to track history when commits are broken into unique bug/feature commits.

File: sdk/core/azure-core/src/http/telemetry_policy.cpp [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    std::string const& componentName,
+    std::string const& componentVersion,
+    std::string const& applicationId)
+{
+  // Spec: https://azure.github.io/azure-sdk/general_azurecore.html#telemetry-policy
+  std::ostringstream telemetryId;
+
+  if (!applicationId.empty())
+  {
+    telemetryId << applicationId.substr(0, 24) << " ";
```
Comment (RickWinter) [inline sdk/core/azure-core/src/http/telemetry_policy.cpp:22]: The applicationId might have whitespace at the end.  Trim the whitespace.

Comment (antkmsft) [reply sdk/core/azure-core/src/http/telemetry_policy.cpp]: Resolved.

File: sdk/core/azure-core/inc/http/policy.hpp (hunk 1 above)
Comment (antkmsft) [reply sdk/core/azure-core/inc/http/policy.hpp:144]: Per our standup conversation, I am not doing this. The behavior is copied from .NET SDK, and the Go factory you mentioned does create the policy for a specific module, this can be done in C++, but should not be done at this (core) level.

File: sdk/core/azure-core/inc/credentials/policy/policies.hpp (hunk 3 above)
Comment (antkmsft) [reply sdk/core/azure-core/inc/credentials/policy/policies.hpp:27]: Yes, it ended up being more than I realized. In the future, I'll be splitting into smaller PRs.

File: sdk/storage/inc/common/common_headers_request_policy.hpp (hunk 2 above)
Comment (antkmsft) [reply sdk/storage/inc/common/common_headers_request_policy.hpp:15]: Not doing anything in this PR, as there is a clarification, and it is not clear whether we will need to add new headers, in which case we may need to do the rename again, please reactivate this comment if you think we should rename it now.

File: sdk/core/azure-core/inc/credentials/policy/policies.hpp (hunk 3 above)
Comment (antkmsft) [reply sdk/core/azure-core/inc/credentials/policy/policies.hpp:27]: Resolving.

File: sdk/core/azure-core/src/http/telemetry_policy.cpp [hunk 5]
```diff
@@ -0,0 +1,143 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#include <http/policy.hpp>
+
+#include <sstream>
```
Comment (RickWinter) [inline sdk/core/azure-core/src/http/telemetry_policy.cpp:6]: This OSVersion helper should be in utils.hpp and utils.cpp within Azure::Core::Utilities namespace

File: sdk/storage/inc/common/shared_key_policy.hpp [hunk 6]
```diff
@@ -17,7 +17,10 @@ namespace Azure { namespace Storage {
 
     ~SharedKeyPolicy() override {}
 
-    HttpPolicy* Clone() const override { return new SharedKeyPolicy(m_credential); }
+    std::unique_ptr<HttpPolicy> Clone() const override
```
Comment (RickWinter) [inline sdk/storage/inc/common/shared_key_policy.hpp:20]: Ensure the storage team reviews and signs off on this change.

File: sdk/core/azure-core/src/http/telemetry_policy.cpp (hunk 5 above)
Comment (antkmsft) [reply sdk/core/azure-core/src/http/telemetry_policy.cpp:6]: I disagree - I would keep it here until there's another place that needs to use this function.

Comment (antkmsft) [reply sdk/core/azure-core/src/http/telemetry_policy.cpp:6]: And there's currently no utils.cpp / hpp. That would need to be an internal header, because we don't want to ship it as public API. What would be the other place that uses it? I don't think that we should move this function there at this moment.

File: sdk/core/azure-core/src/http/telemetry_policy.cpp [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+            {
+              osVersionInfo << (first ? "" : " ")
+                            << std::string(valueBuf, valueBuf + (valueBufSize - 1));
+              first = false;
+            }
+          }
+        }
+      }
+
+      RegCloseKey(regKey);
```
Comment (RickWinter) [inline sdk/core/azure-core/src/http/telemetry_policy.cpp:56]: We only need to close the key if we successfully opened the key.

Comment (antkmsft) [reply sdk/core/azure-core/src/http/telemetry_policy.cpp:56]: That's what we do - it is inside the `if(OpenKey)`

File: sdk/core/azure-core/src/http/telemetry_policy.cpp [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+              "SOFTWARE\\Microsoft\\Windows NT\\CurrentVersion",
+              0,
+              KEY_READ,
+              &regKey)
+          == ERROR_SUCCESS)
+      {
+        regKeyOpened = true;
+
+        auto first = true;
+        char const* regValues[]{
```
Comment (RickWinter) [inline sdk/core/azure-core/src/http/telemetry_policy.cpp:36]: The array of values we look for should be "static constexpr"

File: sdk/core/azure-core/src/http/telemetry_policy.cpp (hunk 7 above)
Comment (antkmsft) [reply sdk/core/azure-core/src/http/telemetry_policy.cpp:56]: Oh, it wasn't! I'll fix.

File: sdk/core/azure-core/src/http/telemetry_policy.cpp [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+  return osVersionInfo.str();
+}
+
+} // namespace
+
+#else
+
+#include <sys/utsname.h>
+
+namespace {
```
Comment (RickWinter) [inline sdk/core/azure-core/src/http/telemetry_policy.cpp:78]: If GetOSVersion doesn't move to Utilities namespace then it needs to be in Azure::Core::Http::Details namespace

File: sdk/core/azure-core/src/http/telemetry_policy.cpp (hunk 8 above)
Comment (antkmsft) [reply sdk/core/azure-core/src/http/telemetry_policy.cpp]: Done.

File: sdk/core/azure-core/src/http/telemetry_policy.cpp (hunk 9 above)
Comment (antkmsft) [reply sdk/core/azure-core/src/http/telemetry_policy.cpp:78]: it is in unnamed namespace. Detail namespace is only for headers.

Comment (antkmsft) [reply sdk/core/azure-core/src/http/telemetry_policy.cpp:78]: Because you can't (shouldn't) do unnamed namespaces in headers - they don't have the intended effect, and at the same time you need to make things available in the headers, but to your code. If the declaration is in the unnamed namespace in .cpp file, it is perfectly accessible the the code in the .cpp file, and only to that code, and the symbolas are not exported by linker (i.e. it is similar to `static` functions in C)

**Comment by antkmsft** [comment]
In teams conversation @RickWinter agreed with the current changes, but only asked me to add tests. I added them. I am pinging him again to unblock the PR.
